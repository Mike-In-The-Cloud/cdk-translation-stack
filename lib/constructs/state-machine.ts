import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as stepfunctions from 'aws-cdk-lib/aws-stepfunctions';
import * as tasks from 'aws-cdk-lib/aws-stepfunctions-tasks';
import * as lambda from 'aws-cdk-lib/aws-lambda'; // Import lambda types
import * as iam from 'aws-cdk-lib/aws-iam'; // Import iam types

export interface StateMachineConstructProps {
  // Accept individual functions instead of the whole construct
  readonly documentProcessorLambda: lambda.IFunction;
  readonly translationProcessorLambda: lambda.IFunction;
  readonly documentCombinerLambda: lambda.IFunction;
  readonly notificationSenderLambda?: lambda.IFunction; // Optional
  // REMOVE: readonly role?: iam.IRole; // Role is now handled externally
}

export class StateMachineConstruct extends Construct {
  // Expose the definition, not the state machine resource
  public readonly definition: stepfunctions.IChainable;
  // REMOVE: public readonly stateMachine: stepfunctions.StateMachine;

  constructor(scope: Construct, id: string, props: StateMachineConstructProps) {
    super(scope, id);

    // Use the specific lambda props
    const {
      documentProcessorLambda,
      translationProcessorLambda,
      documentCombinerLambda,
      notificationSenderLambda
    } = props;

    // --- Step Functions workflow definition ---
    const validateDocument = new tasks.LambdaInvoke(this, 'ValidateDocument', {
      lambdaFunction: documentProcessorLambda, // Use prop
      outputPath: '$.Payload',
    });

    const translateSection = new tasks.LambdaInvoke(this, 'TranslateSection', {
      lambdaFunction: translationProcessorLambda, // Use prop
      inputPath: '$',
      outputPath: '$.Payload',
    });

    const combineDocuments = new tasks.LambdaInvoke(this, 'CombineDocuments', {
      lambdaFunction: documentCombinerLambda, // Use prop
      inputPath: '$',
      outputPath: '$.Payload',
    });

    // Map state to process sections in parallel
    const mapState = new stepfunctions.Map(this, 'MapState', {
        maxConcurrency: 5, // Example concurrency limit
        itemsPath: stepfunctions.JsonPath.stringAt('$.documentSections'), // Assumes validateDocument outputs { ..., documentSections: [...] }
        inputPath: '$', // Pass the full input to the Map state iterations
        parameters: { // Define input for each iteration (translateSection lambda)
            "bucket.$'": '$.bucket', // Pass bucket from original input
            "key.$'": '$.key', // Pass key from original input
            "sourceLanguage.$'": '$.sourceLanguage', // Pass sourceLanguage from original input
            "targetLanguage.$'": '$.targetLanguage', // Pass targetLanguage from original input
            "section.$'": stepfunctions.JsonPath.stringAt('$$.Map.Item.Value') // Pass the current item from itemsPath
        },
        resultPath: '$.translatedSections' // Collect results under this key
    });
    mapState.iterator(translateSection); // Define the task to run for each item


    let workflowDefinition = stepfunctions.Chain.start(validateDocument)
      .next(mapState) // Use Map state instead of Parallel
      .next(combineDocuments);

    // --- Handle Optional Notification Sender in Workflow ---
    if (notificationSenderLambda) { // Check the specific optional prop
      const sendNotification = new tasks.LambdaInvoke(this, 'SendNotification', {
        lambdaFunction: notificationSenderLambda, // Use prop
        inputPath: '$', // Pass the output of combineDocuments
      });
      // Chain notification after combineDocuments
      // Make sure combineDocuments returns necessary info for notification
      workflowDefinition = workflowDefinition.next(sendNotification);
    }

    // Assign the final definition chain to the public property
    this.definition = workflowDefinition;

    // REMOVE: State Machine creation logic
    // const stateMachineName = `${cdk.Aws.STACK_NAME}-TranslationWorkflow`;
    // this.stateMachine = new stepfunctions.StateMachine(this, 'TranslationStateMachine', {
    //   stateMachineName: stateMachineName,
    //   definitionBody: stepfunctions.DefinitionBody.fromChainable(workflowDefinition),
    //   role: props.role, // Use passed role or let it create one
    // });
  }
}