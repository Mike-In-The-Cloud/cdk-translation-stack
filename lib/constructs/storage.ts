import * as cdk from 'aws-cdk-lib';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as opensearch from 'aws-cdk-lib/aws-opensearchserverless';
import { Construct } from 'constructs';

export interface StorageConstructProps {
  tmxBucketName: string;
  documentsBucketName: string;
  openSearchCollectionName: string;
  // Add other relevant config like removal policies if needed
}

export class StorageConstruct extends Construct {
  public readonly tmxBucket: s3.Bucket;
  public readonly documentsBucket: s3.Bucket;
  public readonly translationCollection: opensearch.CfnCollection;

  constructor(scope: Construct, id: string, props: StorageConstructProps) {
    super(scope, id);

    this.tmxBucket = new s3.Bucket(this, 'TMXBucket', {
      bucketName: props.tmxBucketName,
      versioned: true,
      encryption: s3.BucketEncryption.S3_MANAGED,
      removalPolicy: cdk.RemovalPolicy.DESTROY, // Consider making this configurable
    });

    this.documentsBucket = new s3.Bucket(this, 'DocumentsBucket', {
      bucketName: props.documentsBucketName,
      versioned: true,
      encryption: s3.BucketEncryption.S3_MANAGED,
      removalPolicy: cdk.RemovalPolicy.DESTROY, // Consider making this configurable
    });

    this.translationCollection = new opensearch.CfnCollection(this, 'TranslationCollection', {
      name: props.openSearchCollectionName,
      description: 'Collection for storing translation embeddings',
      type: 'VECTORSEARCH'
    });
  }
}