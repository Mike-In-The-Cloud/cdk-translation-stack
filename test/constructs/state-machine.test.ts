import * as cdk from 'aws-cdk-lib';
import { Template, Match } from 'aws-cdk-lib/assertions';
import { Construct } from 'constructs';
import { StateMachineConstruct } from '../../lib/constructs/state-machine';
import * as lambda from 'aws-cdk-lib/aws-lambda';

describe('StateMachineConstruct', () => {
    let stack: cdk.Stack;
    let template: Template;
    // Use real lambda functions
    let docProcessorLambda: lambda.Function;
    let translationProcessorLambda: lambda.Function;
    let docCombinerLambda: lambda.Function;
    let notifierLambda: lambda.Function | undefined;

    const setup = (withNotifier: boolean) => {
        stack = new cdk.Stack();
        // Create real lambda functions
        docProcessorLambda = new lambda.Function(stack, 'TestDocProcessor', {
            runtime: lambda.Runtime.NODEJS_LATEST,
            handler: 'index.handler',
            code: lambda.Code.fromInline('exports.handler = async () => {};'),
        });
        translationProcessorLambda = new lambda.Function(stack, 'TestTranslateProcessor', {
            runtime: lambda.Runtime.NODEJS_LATEST,
            handler: 'index.handler',
            code: lambda.Code.fromInline('exports.handler = async () => {};'),
        });
        docCombinerLambda = new lambda.Function(stack, 'TestDocCombiner', {
            runtime: lambda.Runtime.NODEJS_LATEST,
            handler: 'index.handler',
            code: lambda.Code.fromInline('exports.handler = async () => {};'),
        });
        notifierLambda = withNotifier ? new lambda.Function(stack, 'TestNotifier', {
            runtime: lambda.Runtime.NODEJS_LATEST,
            handler: 'index.handler',
            code: lambda.Code.fromInline('exports.handler = async () => {};'),
        }) : undefined;

        new StateMachineConstruct(stack, 'MyTestStateMachine', {
            // Pass real lambda functions
            documentProcessorLambda: docProcessorLambda,
            translationProcessorLambda: translationProcessorLambda,
            documentCombinerLambda: docCombinerLambda,
            notificationSenderLambda: notifierLambda,
        });
        template = Template.fromStack(stack);
    };

    // Test case with Notification Sender enabled
    describe('With Notification Sender', () => {
        beforeAll(() => {
            setup(true);
        });

        test('Snapshot Test', () => {
            // Snapshot will change
            expect(template.toJSON()).toMatchSnapshot();
        });

        test('Creates State Machine', () => {
            template.resourceCountIs('AWS::StepFunctions::StateMachine', 1);
            template.hasResourceProperties('AWS::StepFunctions::StateMachine', {
                RoleArn: Match.objectLike({ 'Fn::GetAtt': [Match.stringLikeRegexp('MyTestStateMachineTranslationStateMachineRole.*'), 'Arn'] })
            });
        });

        test('References Correct Lambda Functions', () => {
            const docProcessorLogicalId = stack.getLogicalId(docProcessorLambda.node.defaultChild as cdk.CfnElement);
            const translatorLogicalId = stack.getLogicalId(translationProcessorLambda.node.defaultChild as cdk.CfnElement);
            const combinerLogicalId = stack.getLogicalId(docCombinerLambda.node.defaultChild as cdk.CfnElement);
            const notifierLogicalId = notifierLambda ? stack.getLogicalId(notifierLambda.node.defaultChild as cdk.CfnElement) : null;

            template.hasResourceProperties('AWS::StepFunctions::StateMachine', {
                 DefinitionString: Match.objectLike({
                     'Fn::Join': [
                        "", // Separator
                        // Check that the array part contains the Fn::GetAtt for each expected Lambda ARN
                        Match.arrayWith([
                            Match.objectLike({ "Fn::GetAtt": [ docProcessorLogicalId, "Arn" ] }),
                            Match.objectLike({ "Fn::GetAtt": [ translatorLogicalId, "Arn" ] }),
                            Match.objectLike({ "Fn::GetAtt": [ combinerLogicalId, "Arn" ] }),
                            // Conditionally check for the notifier ARN object
                            ...(notifierLogicalId ?
                                [Match.objectLike({ "Fn::GetAtt": [ notifierLogicalId, "Arn" ] })]
                                : [] // If no notifier, don't require this object
                            )
                        ])
                    ]
                 })
            });
        });
    });

    // Test case with Notification Sender disabled
    describe('Without Notification Sender', () => {
        beforeAll(() => {
            setup(false);
        });

        test('Snapshot Test', () => {
            // Snapshot will change
            expect(template.toJSON()).toMatchSnapshot();
        });

        test('Creates State Machine', () => {
            template.resourceCountIs('AWS::StepFunctions::StateMachine', 1);
            template.hasResourceProperties('AWS::StepFunctions::StateMachine', {
                RoleArn: Match.objectLike({ 'Fn::GetAtt': [Match.stringLikeRegexp('MyTestStateMachineTranslationStateMachineRole.*'), 'Arn'] })
            });
        });

        test('Does NOT Reference Notifier Lambda Function (when disabled)', () => {
            const stateMachineResource = template.findResources('AWS::StepFunctions::StateMachine');
            const definitionString = stateMachineResource[Object.keys(stateMachineResource)[0]].Properties.DefinitionString['Fn::Join'][1].join('');
            // Get the logical ID of the Doc Processor to ensure we don't accidentally match part of it
            const docProcessorLogicalId = stack.getLogicalId(docProcessorLambda.node.defaultChild as cdk.CfnElement);

            // Check that the definition string does NOT contain the ARN structure for a function named TestNotifier
            // (using the ID we gave it in setup)
            expect(definitionString).not.toMatch(new RegExp(`.*:function:TestNotifier.*`));
        });
    });
});