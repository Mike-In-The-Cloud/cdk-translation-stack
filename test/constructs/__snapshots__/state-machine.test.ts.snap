// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StateMachineConstruct With Notification Sender Snapshot Test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "MyTestStateMachineTranslationStateMachine8200E6A2": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "MyTestStateMachineTranslationStateMachineRoleDefaultPolicyA1138B84",
        "MyTestStateMachineTranslationStateMachineRole92DFCD58",
      ],
      "Properties": {
        "DefinitionString": {
          "Fn::Join": [
            "",
            [
              "{"StartAt":"ValidateDocument","States":{"ValidateDocument":{"Next":"MapState","Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Type":"Task","OutputPath":"$.Payload","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke","Parameters":{"FunctionName":"",
              {
                "Fn::GetAtt": [
                  "TestDocProcessor7C5BD33D",
                  "Arn",
                ],
              },
              "","Payload.$":"$"}},"MapState":{"Type":"Map","ResultPath":"$.translatedSections","Next":"CombineDocuments","InputPath":"$","Parameters":{"bucket.$'":"$.bucket","key.$'":"$.key","sourceLanguage.$'":"$.sourceLanguage","targetLanguage.$'":"$.targetLanguage","section.$'.$":"$$.Map.Item.Value"},"ItemsPath":"$.documentSections","MaxConcurrency":5,"Iterator":{"StartAt":"TranslateSection","States":{"TranslateSection":{"End":true,"Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Type":"Task","InputPath":"$","OutputPath":"$.Payload","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke","Parameters":{"FunctionName":"",
              {
                "Fn::GetAtt": [
                  "TestTranslateProcessor223610CC",
                  "Arn",
                ],
              },
              "","Payload.$":"$"}}}}},"CombineDocuments":{"Next":"SendNotification","Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Type":"Task","InputPath":"$","OutputPath":"$.Payload","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke","Parameters":{"FunctionName":"",
              {
                "Fn::GetAtt": [
                  "TestDocCombiner42B1785B",
                  "Arn",
                ],
              },
              "","Payload.$":"$"}},"SendNotification":{"End":true,"Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Type":"Task","InputPath":"$","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke","Parameters":{"FunctionName":"",
              {
                "Fn::GetAtt": [
                  "TestNotifierAF36EF8E",
                  "Arn",
                ],
              },
              "","Payload.$":"$"}}}}",
            ],
          ],
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "MyTestStateMachineTranslationStateMachineRole92DFCD58",
            "Arn",
          ],
        },
        "StateMachineName": {
          "Fn::Join": [
            "",
            [
              {
                "Ref": "AWS::StackName",
              },
              "-TranslationWorkflow",
            ],
          ],
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
      "UpdateReplacePolicy": "Delete",
    },
    "MyTestStateMachineTranslationStateMachineRole92DFCD58": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "states.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "MyTestStateMachineTranslationStateMachineRoleDefaultPolicyA1138B84": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "TestDocProcessor7C5BD33D",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "TestDocProcessor7C5BD33D",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "TestDocCombiner42B1785B",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "TestDocCombiner42B1785B",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "TestNotifierAF36EF8E",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "TestNotifierAF36EF8E",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "TestTranslateProcessor223610CC",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "TestTranslateProcessor223610CC",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MyTestStateMachineTranslationStateMachineRoleDefaultPolicyA1138B84",
        "Roles": [
          {
            "Ref": "MyTestStateMachineTranslationStateMachineRole92DFCD58",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestDocCombiner42B1785B": {
      "DependsOn": [
        "TestDocCombinerServiceRole55047213",
      ],
      "Properties": {
        "Code": {
          "ZipFile": "exports.handler = async () => {};",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "TestDocCombinerServiceRole55047213",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestDocCombinerServiceRole55047213": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestDocProcessor7C5BD33D": {
      "DependsOn": [
        "TestDocProcessorServiceRole22185984",
      ],
      "Properties": {
        "Code": {
          "ZipFile": "exports.handler = async () => {};",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "TestDocProcessorServiceRole22185984",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestDocProcessorServiceRole22185984": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestNotifierAF36EF8E": {
      "DependsOn": [
        "TestNotifierServiceRole1565E067",
      ],
      "Properties": {
        "Code": {
          "ZipFile": "exports.handler = async () => {};",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "TestNotifierServiceRole1565E067",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestNotifierServiceRole1565E067": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestTranslateProcessor223610CC": {
      "DependsOn": [
        "TestTranslateProcessorServiceRoleEE75129D",
      ],
      "Properties": {
        "Code": {
          "ZipFile": "exports.handler = async () => {};",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "TestTranslateProcessorServiceRoleEE75129D",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestTranslateProcessorServiceRoleEE75129D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`StateMachineConstruct Without Notification Sender Snapshot Test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "MyTestStateMachineTranslationStateMachine8200E6A2": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "MyTestStateMachineTranslationStateMachineRoleDefaultPolicyA1138B84",
        "MyTestStateMachineTranslationStateMachineRole92DFCD58",
      ],
      "Properties": {
        "DefinitionString": {
          "Fn::Join": [
            "",
            [
              "{"StartAt":"ValidateDocument","States":{"ValidateDocument":{"Next":"MapState","Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Type":"Task","OutputPath":"$.Payload","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke","Parameters":{"FunctionName":"",
              {
                "Fn::GetAtt": [
                  "TestDocProcessor7C5BD33D",
                  "Arn",
                ],
              },
              "","Payload.$":"$"}},"MapState":{"Type":"Map","ResultPath":"$.translatedSections","Next":"CombineDocuments","InputPath":"$","Parameters":{"bucket.$'":"$.bucket","key.$'":"$.key","sourceLanguage.$'":"$.sourceLanguage","targetLanguage.$'":"$.targetLanguage","section.$'.$":"$$.Map.Item.Value"},"ItemsPath":"$.documentSections","MaxConcurrency":5,"Iterator":{"StartAt":"TranslateSection","States":{"TranslateSection":{"End":true,"Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Type":"Task","InputPath":"$","OutputPath":"$.Payload","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke","Parameters":{"FunctionName":"",
              {
                "Fn::GetAtt": [
                  "TestTranslateProcessor223610CC",
                  "Arn",
                ],
              },
              "","Payload.$":"$"}}}}},"CombineDocuments":{"End":true,"Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Type":"Task","InputPath":"$","OutputPath":"$.Payload","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke","Parameters":{"FunctionName":"",
              {
                "Fn::GetAtt": [
                  "TestDocCombiner42B1785B",
                  "Arn",
                ],
              },
              "","Payload.$":"$"}}}}",
            ],
          ],
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "MyTestStateMachineTranslationStateMachineRole92DFCD58",
            "Arn",
          ],
        },
        "StateMachineName": {
          "Fn::Join": [
            "",
            [
              {
                "Ref": "AWS::StackName",
              },
              "-TranslationWorkflow",
            ],
          ],
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
      "UpdateReplacePolicy": "Delete",
    },
    "MyTestStateMachineTranslationStateMachineRole92DFCD58": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "states.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "MyTestStateMachineTranslationStateMachineRoleDefaultPolicyA1138B84": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "TestDocProcessor7C5BD33D",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "TestDocProcessor7C5BD33D",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "TestDocCombiner42B1785B",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "TestDocCombiner42B1785B",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "TestTranslateProcessor223610CC",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "TestTranslateProcessor223610CC",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MyTestStateMachineTranslationStateMachineRoleDefaultPolicyA1138B84",
        "Roles": [
          {
            "Ref": "MyTestStateMachineTranslationStateMachineRole92DFCD58",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestDocCombiner42B1785B": {
      "DependsOn": [
        "TestDocCombinerServiceRole55047213",
      ],
      "Properties": {
        "Code": {
          "ZipFile": "exports.handler = async () => {};",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "TestDocCombinerServiceRole55047213",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestDocCombinerServiceRole55047213": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestDocProcessor7C5BD33D": {
      "DependsOn": [
        "TestDocProcessorServiceRole22185984",
      ],
      "Properties": {
        "Code": {
          "ZipFile": "exports.handler = async () => {};",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "TestDocProcessorServiceRole22185984",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestDocProcessorServiceRole22185984": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestTranslateProcessor223610CC": {
      "DependsOn": [
        "TestTranslateProcessorServiceRoleEE75129D",
      ],
      "Properties": {
        "Code": {
          "ZipFile": "exports.handler = async () => {};",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "TestTranslateProcessorServiceRoleEE75129D",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestTranslateProcessorServiceRoleEE75129D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
